service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureControlledStorageResourceServiceTest.java

```diff
@@ -0,0 +1,273 @@
+import bio.terra.workspace.service.job.JobService;
+import bio.terra.workspace.service.resource.controlled.ControlledResourceService;
+import bio.terra.workspace.service.resource.controlled.cloud.azure.storage.ControlledAzureStorageResource;
+import bio.terra.workspace.service.resource.controlled.cloud.azure.storageContainer.ControlledAzureStorageContainerResource;
+import bio.terra.workspace.service.resource.controlled.model.AccessScopeType;
+import bio.terra.workspace.service.resource.controlled.model.ManagedByType;
+import bio.terra.workspace.service.resource.controlled.model.PrivateResourceState;
+import bio.terra.workspace.service.resource.model.CloningInstructions;
+import bio.terra.workspace.service.spendprofile.SpendConnectedTestUtils;
+import bio.terra.workspace.service.workspace.Alpha1Service;
+import bio.terra.workspace.service.workspace.WorkspaceService;
+import bio.terra.workspace.service.workspace.model.AzureCloudContext;
+import bio.terra.workspace.service.workspace.model.Workspace;
+import bio.terra.workspace.service.workspace.model.WorkspaceStage;
+import com.azure.storage.blob.sas.BlobContainerSasPermission;
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.util.List;
+import java.util.UUID;
+import java.util.regex.Pattern;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+import org.springframework.beans.factory.annotation.Autowired;
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+public class AzureControlledStorageResourceServiceTest extends BaseAzureTest {
+
+  private static Workspace reusableWorkspace;
+  private Workspace workspace;
+  private UserAccessUtils.TestUser workspaceOwner;
+  private UUID storageContainerId;
+  private final OffsetDateTime startTime = OffsetDateTime.now().minusMinutes(15);
+  private final OffsetDateTime expiryTime = OffsetDateTime.now().plusMinutes(60);
+
+  @Autowired private Alpha1Service alpha1Service;
+  @Autowired private ControlledResourceService controlledResourceService;
+  @Autowired private AzureControlledStorageResourceService azureControlledStorageResourceService;
+  @Autowired private WorkspaceService workspaceService;
+  @Autowired private SamService samService;
+  @Autowired private JobService jobService;
+  @Autowired private UserAccessUtils userAccessUtils;
+  @Autowired private SpendConnectedTestUtils spendUtils;
+  @Autowired private FeatureConfiguration features;
+  @Autowired private AzureTestConfiguration azureConfiguration;
+
+  /**
+   * Retrieve or create a workspace ready for controlled Azure storage resources.
+   *
+   * <p>Reusing a workspace saves time between tests.
+   *
+   * @param user
+   */
+  private Workspace reusableWorkspace(UserAccessUtils.TestUser user) throws Exception {
```

Comment:
I think you could use the `BeforeAll` annotation to create this instance that is then shared across test cases.

Reply:
Good call, that's exactly what I want. I used `BeforeEach`, `AfterEach`, and `AfterAll` and completely forgot about `BeforeAll` 🤦